qupulse/hardware/awgs/base.py

```diff
@@ -226,39 +222,24 @@ def __init__(self, program: AllowedProgramTypes,
 
         self._sample_rate = sample_rate
         
-        self._program_type = program_type
         self._program = program
         
-        if program_type == _ProgramType.Linspace:
-            self._transformed_commands = self._transform_linspace_commands(to_increment_commands(program))
-        
         if waveforms is None:
-            if program_type is _ProgramType.Loop:
-                    waveforms = OrderedDict((node.waveform, None)
-                                        for node in program.get_depth_first_iterator() if node.is_leaf()).keys()
-            elif program_type is _ProgramType.Linspace:
-                    #not so clean
-                    #TODO: also marker handling not optimal
-                    waveforms = OrderedDict((command.waveform, None)
-                                        for command in self._transformed_commands if isinstance(command,Play)).keys()
-            else:
-                raise NotImplementedError()
-                    
+            #!!! this formulation is also unfortunate as the channel transformations are
+            # not applied to the waveforms but only to 'Set'/'Increment'-like commands in LSB
+            waveforms_dict = program.get_waveforms_dict(channels,self._channel_transformations())
+            waveforms = waveforms_dict.keys()
         if waveforms:
-            self._waveforms = OrderedDict(zip(waveforms, self._sample_waveforms(waveforms)))
+            self._waveforms = OrderedDict(zip(waveforms,self._sample_waveforms(waveforms)))
```

Comment:
We should stick to whitespace after comma as it is used by most if not all style guides.